text/0000-quick-debug-macro.md

```diff
@@ -0,0 +1,177 @@
+To increase the utility of the macro, it acts as a pass-through function on the expression by simply printing it and then yielding it. On release builds, the macro is the identity function - thus, the macro can be used in release builds while hurting performance while also helping to debug the program.
+
+## Why not use the `log` crate?
+
+While the `log` crate offers a lot of utility, it first has to be used with `extern crate log;`. A logger then has to be set up before expressions can be logged. It is therefore not suitable for introducing newcommers to the language.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## On debug builds:
+
+First, some preliminaries:
+
+```rust
+#[derive(Debug)] // The type of expr in dbg!(expr) must be Debug.
+struct Point {
+    x: usize,
+    y: usize,
+}
+```
+
+With the following example (which most newcomers will benefit from):
+
+```rust
+fn main() {
+    dbg!(Point {
+        x: 1,
+        y: 2,
+    });
+
+    let p = Point {
+        x: 4,
+        y: 5,
+    };
+    dbg!(p);
+}
+```
+
+The program will print the points to `STDERR` as:
+
+```
+[DEBUGGING, src/main.rs:1:4]:
+=> Point{x: 1, y: 2,} = Point {
+    x: 1,
+    y: 2
+}
+[DEBUGGING, src/main.rs:7:4]:
+=> p = Point {
+    x: 4,
+    y: 5
+}
+```
+
+Here, `7:4` is the line and the column.
+
+You may also save the debugged value to a variable or use it in an expression
+since the debugging is pass-through:
+
+```rust
+fn main() {
+    let x = dbg!(1 + 2);
+    let y = dbg!(x + 1) + dbg!(3);
+    dbg!(y);
+}
+```
+
+This prints the following to `STDERR`:
+
+```
+[DEBUGGING, src/main.rs:1:12]:
+=> 1 + 2 = 3
+[DEBUGGING, src/main.rs:2:12]:
+=> x + 1 = 4
+[DEBUGGING, src/main.rs:2:26]:
+=> 3 = 3
+[DEBUGGING, src/main.rs:3:4]:
+=> y = 7
+```
+
+This way of using the macro will mostly benefit existing Rust programmers.
+
+## On release builds:
+
+The same examples above will print nothing to `STDERR` and will instead simply
+evaluate the expressions.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `dbg!` macro will be implemented as:
+
+```rust
+macro_rules! dbg {
+    ($val: expr) => {
+        {
+            let tmp = $val;
+            if cfg!(debug_assertions) {
+                eprintln!("[DEBUGGING, {}:{}:{}]:\n=> {} = {:#?}",
+                    file!(), line!(), column!(), stringify!($val), tmp );
+            }
+            tmp
+        }
+    }
+}
+```
+
+Branching on `cfg!(debug_assertions)` means that if the program is built as a
+release build, nothing will be printed, and the result of using the macro on
+an expression is simply the expression itself. In effect the result is applying
+the identity function on the expression, but the call will be inlined away such
+that the overhead is zero.
+
+The file name, line number and column is included for increased utility when included in production quality code. The expression is also stringified, so that
```

Comment:
What does "production quality code" mean?

Reply:
Things ready to be included in programs & libraries run in production... This is arguably vague and subjective. It could be interpreted as things ready for `cargo publish`, but probably not.

Reply:
I would argue that things that should be included in programs & libraries run in production should *never* be writing something like

```
[DEBUGGING, foobar.rs]
some-expr => 15
```

to stderr. If you want people to use your library, let them decide if and how they want to consume the debugging information you produce.

Reply:
This would probably only be included when developing/debugging a library, which the developer of the lib does.

Reply:
The "when included in production quality code" bit doesn't seem out of place if this wouldn't be used in production quality code.

Reply:
Should probably be reworded.

The intent is:
+ In the case of a library: If the library is used in production and you are the developer/contributor to it, and you want to make some changes and debug those, the macro should be useful to you when doing debug builds.
+ In the case of a program: The macro should be useful to you when doing debug builds and you shouldn't have to remove them in `--release`.

Reply:
But how is the production-ness of the library relevant in the first case?

Reply:
Changed `production quality` to `non-trivial` per discussion here and on IRC.